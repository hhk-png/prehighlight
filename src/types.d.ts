export type HighlightPrefixLength = (word: string) => number

export interface HighlightOptions {
  inplace?: boolean
  returnDom?: boolean
  returnDomFragment?: boolean
  returnWholeBody?: boolean
  highlightPrefixLength?: HighlightPrefixLength
}

type HTMLNode = string | Node
