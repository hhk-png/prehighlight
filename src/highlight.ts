import type { HighlightOptions, HighlightPrefixLength, HTMLNode } from './types'
import {
  createDocumentFragment,
  createElement,
  createTextNode,
  getElementsByTagName,
  isSpecificWord,
  normalize,
  skippedTags,
} from './utils'

function highlightPrefixLength(word: string): number {
  if (isSpecificWord(word)) {
    return 1
  }
  else if (word.length <= 2) {
    return 1
  }
  else if (word.length <= 5) {
    return 2
  }
  else {
    return Math.ceil(word.length / 2)
  }
}

function highlightSentence(textNode: Text, calculatePrefixLength: HighlightPrefixLength): Node[] {
  // textContent of textNode must not be null
  const content = textNode.textContent!
  // spaces only
  if (/^\s+$/.test(content)) {
    return [createTextNode.call(textNode.ownerDocument, content)]
  }

  const nodesList: Node[] = []
  const words = content.split(/(\s+|[.,!?;()])/).filter(Boolean)
  for (const word of words) {
    // english word
    if (/^[a-z]+$/i.test(word)) {
      const boldLength = calculatePrefixLength(word)

      const boldPart = createElement.call(textNode.ownerDocument, 'b')
      boldPart.textContent = word.slice(0, boldLength)
      const restPart = createTextNode.call(textNode.ownerDocument, word.slice(boldLength))
      nodesList.push(boldPart, restPart)
    }
    else {
      // punctuation and chinese, etc.
      const punctuation = createTextNode.call(textNode.ownerDocument, word)
      nodesList.push(punctuation)
    }
  }

  return nodesList
}

export function prehighlight(html: HTMLNode, options: HighlightOptions = {}): HTMLNode {
  if (typeof html === 'string' && options.inplace) {
    throw new Error('inplace option only works with Node type')
  }

  const calculatePrefixLength = options.highlightPrefixLength || highlightPrefixLength

  const domParser = new DOMParser()
  // body is <body> tag
  let body!: Node
  if (!options.inplace) {
    if (typeof html === 'string') {
      const doc = domParser.parseFromString(html, 'text/html')
      body = getElementsByTagName.call(doc, 'body')[0]
    }
    else {
      // html is Node type
      const doc = domParser.parseFromString('<!---->', 'text/html')
      body = getElementsByTagName.call(doc, 'body')[0]
      const importedNode = body.ownerDocument!.importNode(html, true)

      if (
        importedNode.nodeType === 1
        && importedNode.nodeName === 'BODY'
      ) {
        body = importedNode
      }
      else if (importedNode.nodeName === 'HTML') {
        body = getElementsByTagName.call(importedNode, 'body')[0]
      }
      else {
        body.appendChild(importedNode)
      }
    }
  }

  const textNodes: Text[] = []
  function traverseDOM(node: Node): void {
    if (node.nodeType === Node.TEXT_NODE) {
      textNodes.push(node as Text)
    }
    else if (node.nodeType === Node.ELEMENT_NODE) {
      if (!skippedTags.has((node as Element).tagName)) {
        node.childNodes.forEach(traverseDOM)
      }
    }
  }
  traverseDOM(options.inplace ? (html as Node) : body)

  // for concurrent highlight
  // Promise.all(textNodes.map(async (textNode) => {
  //   const nodesList = highlightSentence(textNode)
  //   textNode.replaceWith(...nodesList)
  // }))

  textNodes.forEach((textNode) => {
    const highlightedNodes = highlightSentence(textNode as Text, calculatePrefixLength)
    textNode.replaceWith(...highlightedNodes)
  })

  // normalize html
  normalize.call(html)
  body && normalize.call(body)

  if (options.inplace) {
    return html
  }

  if (options.returnDom) {
    if (options.returnDomFragment) {
      const returnedNode = createDocumentFragment.call(body.ownerDocument)

      while (body.firstChild) {
        returnedNode.appendChild(body.firstChild)
      }

      return returnedNode
    }

    return body
  }

  return options.returnWholeBody
    ? (body as HTMLElement).outerHTML
    : (body as HTMLElement).innerHTML
}
