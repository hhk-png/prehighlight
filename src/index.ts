
interface HighlightOptions {
  inplace?: boolean
  returnDom?: boolean
  returnDomFragment?: boolean
  returnWholeBody?: boolean
}

const {
  getElementsByTagName,
  createNodeIterator,
  createDocumentFragment,
} = document

function createTextNodeIterator(node: Node) {
  return createNodeIterator.call(
    node.ownerDocument,
    node,
    NodeFilter.SHOW_TEXT,
    null
  )
}

export function prehighlight(html: string | Node, cfg: HighlightOptions = {}) {

  if (typeof html === 'string' && cfg.inplace) {
    throw new Error('inplace option only works with Node type')
  }

  const domParser = new DOMParser()
  // body is <body> tag
  let body!: Node
  if (typeof html === 'string') {
    const doc = domParser.parseFromString(html, 'text/html')
    body = getElementsByTagName.call(doc, 'body')[0]
  }
  else {
    // html is Node type
    const doc = domParser.parseFromString('<!---->', 'text/html')
    body = getElementsByTagName.call(doc, 'body')[0]
    const importedNode = body.ownerDocument!.importNode(html, true)

    if (
      importedNode.nodeType === 1 &&
      importedNode.nodeName === 'BODY'
    ) {
      body = importedNode
    }
    else if (importedNode.nodeName === 'HTML') {
      body = getElementsByTagName.call(importedNode, 'body')[0]
    }
    else {
      body.appendChild(importedNode)
    }
  }

  const textNodeIterator = createTextNodeIterator(cfg.inplace ? (html as Node) : body)

  let currentTextNode: Node | null
  let temp
  while ((currentTextNode = textNodeIterator!.nextNode())) {
    // TODO
    temp = currentTextNode.textContent
  }


  if (cfg.returnDom) {
    if (cfg.returnDomFragment) {
      const returnedNode = createDocumentFragment.call(body.ownerDocument)

      while (body.firstChild) {
        returnedNode.appendChild(body.firstChild)
      }

      return returnedNode
    }

    return body
  }

  return cfg.returnWholeBody
    ? (body as HTMLElement).outerHTML
    : (body as HTMLElement).innerHTML
}
