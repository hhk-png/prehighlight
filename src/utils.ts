const {
  getElementsByTagName,
  createDocumentFragment,
  createTextNode,
  createElement,
  normalize,
} = document

export {
  createDocumentFragment,
  createElement,
  createTextNode,
  getElementsByTagName,
  normalize,
}

export const specificWords = new Set([
  'the',
  'and',
  'in',
  'on',
  'at',
  'by',
  'with',
  'about',
  'against',
  'between',
  'into',
  'through',
  'during',
  'before',
  'after',
  'above',
  'below',
  'to',
  'from',
  'up',
  'down',
  'over',
  'under',
  'again',
  'further',
  'then',
  'once',
  'here',
  'there',
  'when',
  'where',
  'why',
  'how',
  'all',
  'any',
  'both',
  'each',
  'few',
  'more',
  'most',
  'other',
  'some',
])

export function isSpecificWord(word: string): boolean {
  return specificWords.has(word.toLowerCase())
}

export const skippedTags = new Set([
  'PRE',
  'CODE',
  'SCRIPT',
  'STYLE',
  'TITLE',
  'HEAD',
  'META',
  'LINK',
  'NOSCRIPT',
  'TEMPLATE',
  'SVG',
  'CANVAS',
  'SLOT',
])
