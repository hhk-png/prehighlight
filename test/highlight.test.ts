import { describe, expect, it } from 'vitest'
import { prehighlight } from '../src/index'

describe('prehighlight config', () => {
  it('options.inplace do not work with string', () => {
    const html = '<div>hello world</div>'
    const options = { inplace: true }
    expect(() => prehighlight(html, options)).toThrowError('inplace option only works with Node type')
  })

  it('options.inplace works with dom', () => {
    const html = new DOMParser().parseFromString('<div>hello world</div>', 'text/html')
    const options = { inplace: true }
    const res = prehighlight(html.body.firstChild!, options)
    expect(res).toBeInstanceOf(Node)
    expect((res as Element).innerHTML).toBe('<b>he</b>llo <b>wo</b>rld')
  })

  it('options.returnDom', () => {
    const html = '<div>hello world</div>'
    const options = { returnDom: true }
    expect(prehighlight(html, options)).toBeInstanceOf(Node)
  })

  it('options.returnDom with options.returnDomFragment', () => {
    const html = '<div>hello world</div>'
    const options = { returnDom: true, returnDomFragment: true }
    const res = prehighlight(html, options) as Element
    expect(res).toBeInstanceOf(DocumentFragment)
    expect(res.childNodes[0].childNodes.length).toBe(4)
  })

  it('options.returnWholeDocument', () => {
    const html = '<div>hello world</div>'
    const options = { returnWholeBody: true }
    expect(prehighlight(html, options)).toBe('<body><div><b>he</b>llo <b>wo</b>rld</div></body>')
  })

  it('highlightPrefixLength', () => {
    const html = '<div>hello world</div>'
    const options = { highlightPrefixLength: (_: string) => 1 }
    expect(prehighlight(html, options)).toBe('<div><b>h</b>ello <b>w</b>orld</div>')
  })
})

describe('highlight dom', () => {
  it('highlight dom', () => {
    const dom = new DOMParser().parseFromString('<div>hello world</div>', 'text/html')
    expect(prehighlight(dom.body.firstChild!)).toBe('<div><b>he</b>llo <b>wo</b>rld</div>')
  })

  it('highlight dom with <body>', () => {
    const dom = new DOMParser().parseFromString('<body>hello world</body>', 'text/html')
    expect(prehighlight(dom.body)).toBe('<b>he</b>llo <b>wo</b>rld')
  })

  it('highlight dom with <html>', () => {
    const dom = new DOMParser().parseFromString('<html><body>hello world</body></html>', 'text/html')
    expect(prehighlight(dom.documentElement)).toBe('<b>he</b>llo <b>wo</b>rld')
  })
})

describe('hightlight complex html', () => {
  it('english', () => {
    const html = `    <div>
      <p>The DOM, or Document Object Model</p>
      <p>dangerous</p>
      <p>This is very <i>important</i></p>
    </div>`

    const res = prehighlight(html)
    expect(res).toMatchSnapshot()
  })

  it('highlight chinese', () => {
    const html = `<div>你好，世界！</div>`
    const res = prehighlight(html)
    expect(res).toBe('<div>你好，世界！</div>')
  })

  it('empty tag', () => {
    const html = ` <div></div>`
    const res = prehighlight(html)
    expect(res).toBe('<div></div>')
  })

  it('complex html', () => {
    const html = `  <header>
    <h1>Bionic Reading 演示</h1>
    <p>测试复杂结构下的文本渲染</p>
  </header>

  <nav>
    <a href="#">主页</a>
    <a href="#">博客</a>
    <a href="#">项目</a>
    <a href="#">联系</a>
  </nav>

  <article>
    <h2>引言</h2>
    <p>随着信息爆炸式增长，<strong>高效阅读</strong>变得尤为重要。Bionic Reading 通过突出词语的前缀，帮助读者更快识别内容，提高理解力。</p>

    <h2>段落示例</h2>
    <p>在这个段落中，我们会混合使用<strong>粗体</strong>、<em>斜体</em>、超链接以及引用：</p>
    <blockquote>“这是一个引用示例，用于测试 Bionic Reading 的表现。”</blockquote>
    <p>你可以访问<a href="https://www.example.com">这个链接</a>获取更多信息。</p>

    <h2>图片示例</h2>
    <img src="https://via.placeholder.com/600x200" alt="示意图">

    <h2>代码块</h2>
    <pre><code>function greet(name) {
  return 'Hello, ' + name + '!';
}</code></pre>

    <h2>表格</h2>
    <table>
      <thead>
        <tr>
          <th>项目</th>
          <th>状态</th>
          <th>进度</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>页面设计</td>
          <td>完成</td>
          <td>100%</td>
        </tr>
        <tr>
          <td>功能开发</td>
          <td>进行中</td>
          <td>60%</td>
        </tr>
      </tbody>
    </table>
  </article>

  <footer>
    <p>&copy; 2025 示例网站 - Bionic Reading 测试</p>
  </footer>
  <script>
    console.log('This is a script tag.')
  </script>

  `
    const res = prehighlight(html)
    expect(res).toMatchSnapshot()
  })
})

describe('highlight shadow dom', () => {
  it('opened shadow dom', () => {
    class MyCard extends HTMLElement {
      constructor() {
        super()

        const shadowRoot = this.attachShadow({ mode: 'open' })

        shadowRoot.innerHTML = `
        <style>
          .card {
            border: 2px solid #ccc;
            padding: 10px;
          }
          ::slotted(span) {
            color: blue;
          }
        </style>
        <div class="card">
          <h2><slot name="title"></slot></h2> <!-- slot -->
          <slot></slot> <!-- default slot -->
        </div>
      `
      }
    }

    customElements.define('my-card', MyCard)

    const myCardElement = document.createElement('my-card')
    document.body.appendChild(myCardElement)

    const title = document.createElement('span')
    title.setAttribute('slot', 'title')
    title.textContent = 'My Custom Card Title'
    myCardElement.appendChild(title)

    const content = document.createElement('p')
    content.textContent = 'This is some card content.'
    myCardElement.appendChild(content)

    const res = prehighlight(myCardElement)
    expect(res).toMatchSnapshot('opened shadow dom')
  })

  it('closed shadow dom', () => {
    class MyCard extends HTMLElement {
      constructor() {
        super()

        const shadowRoot = this.attachShadow({ mode: 'closed' })

        shadowRoot.innerHTML = `
        <style>
          .card {
            border: 2px solid #ccc;
            padding: 10px;
          }
          ::slotted(span) {
            color: blue;
          }
        </style>
        <div class="card">
          <h2><slot name="title"></slot></h2> <!-- slot -->
          <slot></slot> <!-- default slot -->
        </div>
      `
      }
    }

    customElements.define('my-card-closed', MyCard)

    const myCardElement = document.createElement('my-card-closed')
    document.body.appendChild(myCardElement)

    const title = document.createElement('span')
    title.setAttribute('slot', 'title')
    title.textContent = 'My Custom Card Title'
    myCardElement.appendChild(title)

    const content = document.createElement('p')
    content.textContent = 'This is some card content.'
    myCardElement.appendChild(content)

    const res = prehighlight(myCardElement)
    expect(res).toMatchSnapshot('closed shadow dom')
  })
})
