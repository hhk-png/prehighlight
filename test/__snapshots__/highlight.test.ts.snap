// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`highlight shadow dom > closed shadow dom > closed shadow dom 1`] = `"<my-card-closed><span slot="title"><b>M</b>y <b>Cus</b>tom <b>Ca</b>rd <b>Ti</b>tle</span><p><b>Th</b>is <b>i</b>s <b>s</b>ome <b>ca</b>rd <b>cont</b>ent.</p></my-card-closed>"`;

exports[`highlight shadow dom > opened shadow dom > opened shadow dom 1`] = `"<my-card><span slot="title"><b>M</b>y <b>Cus</b>tom <b>Ca</b>rd <b>Ti</b>tle</span><p><b>Th</b>is <b>i</b>s <b>s</b>ome <b>ca</b>rd <b>cont</b>ent.</p></my-card>"`;

exports[`hightlight complex html > complex html 1`] = `
"<header>
    <h1><b>Bio</b>nic <b>Read</b>ing 演示</h1>
    <p>测试复杂结构下的文本渲染</p>
  </header>

  <nav>
    <a href="#">主页</a>
    <a href="#">博客</a>
    <a href="#">项目</a>
    <a href="#">联系</a>
  </nav>

  <article>
    <h2>引言</h2>
    <p>随着信息爆炸式增长，<strong>高效阅读</strong>变得尤为重要。Bionic <b>Read</b>ing 通过突出词语的前缀，帮助读者更快识别内容，提高理解力。</p>

    <h2>段落示例</h2>
    <p>在这个段落中，我们会混合使用<strong>粗体</strong>、<em>斜体</em>、超链接以及引用：</p>
    <blockquote>“这是一个引用示例，用于测试 <b>Bio</b>nic <b>Read</b>ing 的表现。”</blockquote>
    <p>你可以访问<a href="https://www.example.com">这个链接</a>获取更多信息。</p>

    <h2>图片示例</h2>
    <img src="https://via.placeholder.com/600x200" alt="示意图">

    <h2>代码块</h2>
    <pre><code>function greet(name) {
  return 'Hello, ' + name + '!';
}</code></pre>

    <h2>表格</h2>
    <table>
      <thead>
        <tr>
          <th>项目</th>
          <th>状态</th>
          <th>进度</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>页面设计</td>
          <td>完成</td>
          <td>100%</td>
        </tr>
        <tr>
          <td>功能开发</td>
          <td>进行中</td>
          <td>60%</td>
        </tr>
      </tbody>
    </table>
  </article>

  <footer>
    <p>&amp;copy; 2025 示例网站 - <b>Bio</b>nic <b>Read</b>ing 测试</p>
  </footer>
  <script>
    console.log('This is a script tag.')
  </script>

  "
`;

exports[`hightlight complex html > english 1`] = `
"<div>
      <p><b>T</b>he <b>DO</b>M, <b>o</b>r <b>Docu</b>ment <b>Obj</b>ect <b>Mo</b>del</p>
      <p><b>dange</b>rous</p>
      <p><b>Th</b>is <b>i</b>s <b>ve</b>ry <i><b>impor</b>tant</i></p>
    </div>"
`;
